crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2407,6 +2407,104 @@ impl<'db> TypeInferenceBuilder<'db> {
                 }
             }
 
+            (
```

Comment:
I might futz with the organization of these and add some comments?

Reply:
cc @carljm regarding the handling of `IntLiteral` without it being a special case

Reply:
The algorithm for inferring binary operations is unfortunately really really complicated in its totality, though. See https://snarky.ca/unravelling-binary-arithmetic-operations-in-python/ for all the gory details!

Reply:
Interesting! Thanks for the context.

Reply:
Yep, @AlexWaygood perfectly summarized what I was talking about in standup, and with lots more useful details, too!!

Reply:
Yeah, so either way we are going to have large match statements, but once we are in the realm of "things typeshed can tell us" (which is all of the cases added in this PR), we should avoid adding new special case match arms and just have a generic version that looks up and "calls" the appropriate dunder methods like Alex suggests. Pretty much we should only have special-case match arms for literal types if there is a possibility we can infer a literal type out of the operation, otherwise we should be falling back to the general case.

Reply:
Makes sense, I was thinking something was wrong here but wasn't aware typeshed does all of this.

Reply:
Oh yeah, we go to great lengths to accurately reflect all the dunders: <https://github.com/python/typeshed/blob/44aa63330b03bdacab731af2333ff9bf70855de3/stdlib/builtins.pyi#L227-L330>

And we have quite extensive testing to check that none are omitted from the stub or have inconsistent signatures with what actually exists at runtime.